Make Chains concretely typed + other breaking changes

devmotion opened the pull request:
This PR contains different breaking changes that were discussed before and/or mentioned in different issues. There are more things that could be included, but I think it might be enough for this PR.

The main changes are:
- The type signature of `Chains` is changed, all fields are concrete types now
- Parameter names and name maps use only symbols internally but users can still provide names, name maps, and perform indexing with strings (should this possibility be removed?)
- The statistics and diagnostics function consistently use lowercase symbols without underscores for the names of the results
- The indexing of a `Chains` object now always just forwards strings and symbols to the second dimension of the underlying AxisArray instead of trying to find other symbols that start with the same name + "[" as well (for vector-valued parameters, e.g.). Instead now a function `namesingroup(chains, name)` (maybe there's a better name?) is provided that returns all corresponding parameter names in the chain and could be used by a user for indexing. The old behaviour of trying to preserve the Chains structure even if only one sample, chain, or iteration is selected is preserved (should this be changed?).
- All `sorted` keyword arguments are removed.
- `set_section` is renamed to `replacenames` since it just performs replacement of parameter names. It now takes replacement pairs instead of a dictionary, in the same way as `replace` in Base does.

cpfiffer commented:
>     * Parameter names and name maps use only symbols internally but users can still provide names, name maps, and perform indexing with strings (should this possibility be removed?)

I think it's fine to leave in string indexing, particularly if you want to query for a specific value, i.e. it's easier to type `b[1]` than `Symbol("b[1]")`.

>     * The indexing of a `Chains` object now always just forwards strings and symbols to the second dimension of the underlying AxisArray instead of trying to find other symbols that start with the same name + "[" as well (for vector-valued parameters, e.g.). Instead now a function `namesingroup(chains, name)` (maybe there's a better name?) is provided that returns all corresponding parameter names in the chain and could be used by a user for indexing. 

This is a monumental change -- I think we should stick this PR into a `dev` branch and basically just do everything there. We can also consider switching to `SturctArray` while we're doing enormous changes like this. Would you be alright with putting this into a` dev` branch so we can leave `master` open for bug fixes and patches?

Additionally, I suggest the function `group(chains, name) = chain[:, namesingroup(chains, name), :]` as a convenience function to retrieve a group of parameters.

> The old behaviour of trying to preserve the Chains structure even if only one sample, chain, or iteration is selected is preserved (should this be changed?).

I _hate_ this behavior, but I don't know what's a better option, because in a lot of cases it's really nice to do something like `plot(chain[1:100, :a, :])` as it returns a chain. I think good conditions for returning an array might be if a single parameter is selected -- I can't think of other good general cases to return an array.

We should return a scalar if the user specifies an iteration, parameter, and chain -- it's pretty obvious that someone in this case just wants a damn number.

I'm good with the other main bullet points you mentioned, but I'm going to do the code review now.

cpfiffer commented on src/stats.jl:
```diff
@@ -435,10 +432,10 @@ function summarystats(chn::Chains;
     # Store everything.
     funs = [mean∘cskip, std∘cskip, sem∘cskip, df_mcse]
-    func_names = [:mean, :std, :naive_se, :mcse]
+    func_names = [:mean, :std, :naivese, :mcse]
```
Shouldn't this have an underscore since it's two words? It's a little hard to read.

devmotion replied:
Yes, I agree, I think in this case the underscore actually improves readability. Alternatively, one could also write a bit more explicitly `:naivestandarderror` (with or without underscore?), `:standarddeviation` (or `:stddev`?), and `:mcstandarderror` or `:montecarlostandarderror`. Mhh but maybe we should keep it shorter :smile:

cpfiffer replied:
Yeah, I think `naive_se` is short and readable. I'm fine with `stddev` or `std`, I'm very familiar with both and I think nearly all R users would agree with `std`.

devmotion commented on src/chains.jl:
```diff
@@ -637,25 +563,25 @@ function pool_chain(c::Chains)
-    # Set new parameter names.
-    params = names(c)
-    new_params = replace(params, d...)
+function replacenames(chains::Chains, old_new::Pair...)
```
Yes, good idea, I thought about this at some point but forgot to add it! We could even think about dropping the `Dict` type annotation - then it would also support other types of dictionaries and named tuples.

devmotion replied:
I've added the convenience function for AbstractDict.

devmotion commented on README.md:
````diff
@@ -76,11 +77,13 @@ val = rand(500,5, 2)
 ```
 
-By default, parameters will be given the name `Parami`, where `i` is the parameter number.
+By default, parameters will be given the name `:parami`, where `i` is the parameter number.
````
I don't have a strong opinion on this. I had the impression a common practice in Julia is to avoid underscores if it doesn't impact readability (e.g., `isconcretetype` etc.) but I'm not sure if it's more readable with or without underscore here.

cpfiffer replied:
I think it's slightly more readable since it's a number on the right hand side, but that's just me.

devmotion replied:
As I said, I don't have a strong opinion on this one, so I'll change it.

devmotion commented:
> I think it's fine to leave in string indexing, particularly if you want to query for a specific value, i.e. it's easier to type b[1] than Symbol("b[1]").

I agree, I had the same impression.

> This is a monumental change -- I think we should stick this PR into a dev branch and basically just do everything there. We can also consider switching to SturctArray while we're doing enormous changes like this. Would you be alright with putting this into a dev branch so we can leave master open for bug fixes and patches?

Sure, we can do that. However, note that due to https://github.com/TuringLang/MCMCChains.jl/pull/193 we have to release a new major breaking release as soon as possible anyway, and the Tables PR changed the indexing behaviour and the removal of keyword arguments in basically the same way.

Actually, that PR removed keyword arguments even more drastically, since also the `append_chains`, `sections`, `digits`, and `showall` options are gone in the construction of dataframes. Should we do the same for all methods? I'm inclined to say yes, since there's actually no advantage for a user in having these keyword arguments. Currently, we subset and transform the chain in every call of these methods depending on the provided keyword arguments, which leads to a lot of redundant code (and also not all methods actually supporting all keyword arguments) and makes it difficult (or almost impossible) for a user to write efficient code by already providing only a subset of a chain. IMO one should rather suggest using the `pool` function instead of `append_chains` and the indexing operations instead of `sections` and `showall`, and drop `digits` alltogether (I don't really understand why it exists at all).

BTW I'm not sure if we should actually replace the current `Chains` setup completely with a StructArray based one. I guess, it could be useful to have both since I don't think that one is always strictly better than the other.

> Additionally, I suggest the function group(chains, name) = chain[:, namesingroup(chains, name), :] as a convenience function to retrieve a group of parameters.

I like this!

> I hate this behavior, but I don't know what's a better option, because in a lot of cases it's really nice to do something like plot(chain[1:100, :a, :]) as it returns a chain. I think good conditions for returning an array might be if a single parameter is selected -- I can't think of other good general cases to return an array.

So you mean `chain[1:100, :a, :]` should actually return an array?

> We should return a scalar if the user specifies an iteration, parameter, and chain -- it's pretty obvious that someone in this case just wants a damn number.

Yes, indeed, that's reasonable.

cpfiffer commented:
> IMO one should rather suggest using the `pool` function instead of `append_chains` and the indexing operations instead of `sections` and `showall`

I'm divided on this one because I really like having default `sections` -- I almost never want to see `lp` or `numerical_error` next to my parameter stats, because those are what I care about. I think this is a debateable preference, though. The best design I think would be to make the user deal with sections, but being able to do `describe(chain, section=[:parameters, :diagnostics])` is a really nice tool.

I'm fine with removing `append_chain` -- that should be done user side unless it's something that we know is pooled, like running top-level summary stats.

> and drop `digits` alltogether (I don't really understand why it exists at all).

My reasoning for this was if you had parameter estimates that are very small when printing info out. We could try to infer the number of significant digits instead, but I think it's a good thing to have for people who work with very small numbers (like me).

> So you mean `chain[1:100, :a, :]` should actually return an array?

If we are removing the ability to use `getindex` on groups, then yes, since `:a` would be a unique parameter.

devmotion commented:
> My reasoning for this was if you had parameter estimates that are very small when printing info out. We could try to infer the number of significant digits instead, but I think it's a good thing to have for people who work with very small numbers (like me).

I see, but IMO it might be more suitable to provide an option for modifying how the statistics are printed when they are actually printed than when they are computed. So I think this should rather be an option of the display method of a ChainsDataFrame than part of every statistics function.

BTW have you seen PrettyTables? I guess maybe one could use that package to print out ChainDataFrames:

```julia
julia> using PrettyTables

julia> pretty_table((mean = rand(10), std = rand(10), naive_se = rand(10)))
┌─────────────────────┬─────────────────────┬─────────────────────┐
│                mean │                 std │            naive_se │
│             Float64 │             Float64 │             Float64 │
├─────────────────────┼─────────────────────┼─────────────────────┤
│     0.6637320214016 │  0.6665236913400143 │  0.9295054083756582 │
│  0.7951800957224184 │  0.5688714199186389 │  0.8325279741445928 │
│ 0.27751139639143574 │  0.5618917529371787 │ 0.07108691136064915 │
│  0.4241231574082598 │   0.515809359903415 │  0.7968483224379328 │
│  0.7789652997978131 │ 0.23821324513654396 │  0.8986365081937169 │
│  0.8979763547057718 │ 0.26562346606611587 │  0.7336567412671315 │
│ 0.41866686387557683 │  0.8444012632857871 │  0.8373881786993129 │
│  0.2331565932303583 │  0.7503745215007509 │ 0.04704294035200873 │
│  0.7961611258634664 │  0.8531175015585246 │ 0.40073552778745847 │
│  0.6564853801348485 │  0.7629574469839067 │  0.3401034998263961 │
└─────────────────────┴─────────────────────┴─────────────────────┘

julia> pretty_table((mean = rand(10), std = rand(10), naive_se = rand(10)); formatter = ft_printf("%5.2f", 2:3))
┌─────────────────────┬─────────┬──────────┐
│                mean │     std │ naive_se │
│             Float64 │ Float64 │  Float64 │
├─────────────────────┼─────────┼──────────┤
│  0.3651274532813491 │    0.29 │     0.85 │
│   0.628570862677025 │    0.76 │     0.89 │
│  0.8685490608257325 │    0.99 │     0.85 │
│ 0.21587848067856052 │    0.80 │     0.90 │
│ 0.16156442643054936 │    0.16 │     0.33 │
│  0.8164557835782176 │    0.43 │     0.24 │
│  0.3651776772369091 │    0.41 │     0.16 │
│ 0.11765101440238257 │    0.13 │     0.97 │
│   0.638144487678705 │    0.49 │     0.60 │
│  0.8751275228685844 │    0.61 │     0.33 │
└─────────────────────┴─────────┴──────────┘
```

devmotion commented:
@cpfiffer: I've changed the indexing behaviour in the following way:
1. If only one parameter is requested, the indices are forwarded to the underlying `AxisArray`, resulting in matrices, vectors, or just a number if a specific iteration, parameter, and a chain are specified
2. If multiple parameters (or an array with one parameter) are specified, a `Chains` object is returned
3. Strings are still supported and converted to symbols appropriately

As an example:
```julia
julia> chn = Chains(rand(10, 4, 5), string.("param[", 1:4, "]"));

julia> namesingroup(chn, "param")
4-element Array{Symbol,1}:
 Symbol("param[1]")
 Symbol("param[2]")
 Symbol("param[3]")
 Symbol("param[4]")

julia> group(chn, "param")
Object of type Chains, with data of type 10×4×5 Array{Float64,3}

Iterations        = 1:10
Thinning interval = 1
Chains            = 1, 2, 3, 4, 5
Samples per chain = 10
parameters        = param[1], param[2], param[3], param[4]

2-element Array{ChainDataFrame,1}

Summary Statistics
  parameters    mean     std  naive_se     mcse      ess    rhat
  ──────────  ──────  ──────  ────────  ───────  ───────  ──────
    param[1]  0.4611  0.3181    0.0450  missing  39.0372  1.0796
    param[2]  0.4531  0.3120    0.0441  missing  28.7014  1.0201
    param[3]  0.5035  0.2996    0.0424  missing  75.2512  0.9949
    param[4]  0.4387  0.2723    0.0385  missing  67.1509  0.9666

Quantiles
  parameters    2.5%   25.0%   50.0%   75.0%   97.5%
  ──────────  ──────  ──────  ──────  ──────  ──────
    param[1]  0.0180  0.1839  0.4241  0.7461  0.9769
    param[2]  0.0200  0.1669  0.3742  0.7250  0.9841
    param[3]  0.0352  0.2088  0.5135  0.7827  0.9387
    param[4]  0.0181  0.2516  0.4098  0.5913  0.9410

julia> chn[2:4]
Object of type Chains, with data of type 3×4×5 Array{Float64,3}

Iterations        = 2:4
Thinning interval = 1
Chains            = 1, 2, 3, 4, 5
Samples per chain = 3
parameters        = param[1], param[2], param[3], param[4]

2-element Array{ChainDataFrame,1}

Summary Statistics
  parameters    mean     std  naive_se     mcse  ess  rhat
  ──────────  ──────  ──────  ────────  ───────  ───  ────
    param[1]  0.4684  0.2994    0.0773  missing  NaN   NaN
    param[2]  0.4183  0.2909    0.0751  missing  NaN   NaN
    param[3]  0.5260  0.3520    0.0909  missing  NaN   NaN
    param[4]  0.5110  0.2644    0.0683  missing  NaN   NaN

Quantiles
  parameters    2.5%   25.0%   50.0%   75.0%   97.5%
  ──────────  ──────  ──────  ──────  ──────  ──────
    param[1]  0.0254  0.2458  0.4400  0.7086  0.8673
    param[2]  0.0254  0.2182  0.3975  0.5685  0.8765
    param[3]  0.0567  0.2090  0.5529  0.8766  0.9500
    param[4]  0.1044  0.3778  0.5196  0.7096  0.9022

julia> chn[["param[1]", "param[4]"]]
Object of type Chains, with data of type 10×2×5 Array{Float64,3}

Iterations        = 1:10
Thinning interval = 1
Chains            = 1, 2, 3, 4, 5
Samples per chain = 10
parameters        = param[1], param[4]

2-element Array{ChainDataFrame,1}

Summary Statistics
  parameters    mean     std  naive_se     mcse      ess    rhat
  ──────────  ──────  ──────  ────────  ───────  ───────  ──────
    param[1]  0.4611  0.3181    0.0450  missing  39.0372  1.0796
    param[4]  0.4387  0.2723    0.0385  missing  67.1509  0.9666

Quantiles
  parameters    2.5%   25.0%   50.0%   75.0%   97.5%
  ──────────  ──────  ──────  ──────  ──────  ──────
    param[1]  0.0180  0.1839  0.4241  0.7461  0.9769
    param[4]  0.0181  0.2516  0.4098  0.5913  0.9410

julia> chn["param[2]"]
2-dimensional AxisArray{Float64,2,...} with axes:
    :iter, 1:1:10
    :chain, 1:5
And data, a 10×5 Array{Float64,2}:
 0.91825    0.350491    0.0206528  0.307175   0.987926
 0.0684399  0.00227728  0.312149   0.868478   0.863866
 0.451117   0.403547    0.137187   0.397523   0.880851
 0.299142   0.110513    0.381041   0.685922   0.412287
 0.144023   0.781663    0.367419   0.985579   0.290788
 0.520222   0.165923    0.670048   0.738037   0.816821
 0.343252   0.354643    0.851009   0.667084   0.952458
 0.475756   0.0883705   0.149303   0.0198101  0.272536
 0.662355   0.97905     0.278098   0.135271   0.0923619
 0.858029   0.169771    0.656442   0.0784744  0.230815

julia> chn[2, "param[1]", :]
1-dimensional AxisArray{Float64,1,...} with axes:
    :chain, 1:5
And data, a 5-element Array{Float64,1}:
 0.38016092191776174
 0.8596835632877888
 0.04265930903160364
 0.016171617913429248
 0.21484140367421456

julia> chn[:, "param[3]", 3]
1-dimensional AxisArray{Float64,1,...} with axes:
    :iter, 1:1:10
And data, a 10-element Array{Float64,1}:
 0.48704512903364683
 0.03092079488300148
 0.30140971477305833
 0.29910985458420547
 0.24419948127245372
 0.21755031282224557
 0.8624359740480712
 0.1500854666880178
 0.04981674313881301
 0.7255330763268573

julia> chn[2, "param[3]", 3]
0.03092079488300148
```

Is that what you had in mind? Do you think that's reasonable and intuitive?

BTW I'd suggest opening a separate PR for removing the keyword arguments and not including them in this PR, it seems this one is already quite big.

cpfiffer commented:
Yeah, I think that's very intuitive. If you've ever worked with `pandas` this has the same feel, which I like. I'm fine with splitting off the keyword arguments stuff into a separate PR. Let me know when you want me to do a second pass on the code review.

> BTW have you seen PrettyTables? I guess maybe one could use that package to print out ChainDataFrames:

I hadn't seen this, no! I wonder how performant it is though. The big reason I removed the dataframe stuff is that the display methods took forever to spin up, but I suspect this is a much lighter dependency.

devmotion commented:
>  Let me know when you want me to do a second pass on the code review.

I think it's ready for a second pass. Thanks!

> I hadn't seen this, no! I wonder how performant it is though. The big reason I removed the dataframe stuff is that the display methods took forever to spin up, but I suspect this is a much lighter dependency.

Yes, it only depends on Crayons, Formatting, Parameters, Reexport, and Tables. I'm just working on a PR that removes the digits keyword and instead uses a IOContext to specify the digits that are printed with PrettyTables.

devmotion commented:
I just merged the latest changes, I guess we're still waiting for @goedman's comments?

devmotion commented:
Resolved some conflicts due to some recent updates to MCMCChains. It would be great if we could get this PR in if everyone agrees with the changes. DynamicPPL was just updated some days ago and now uses only symbols internally, so it would be good to get rid of all the internal conversions from and to strings in MCMCChains as well.

cpfiffer commented:
I'm fine with merging if you are.

devmotion commented:
I'm fine with it :+1:
